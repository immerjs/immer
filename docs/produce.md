---
id: produce
title: Using produce
---

<div id="codefund"><!-- fallback content --></div>

<details>
    <summary style="color: white; background:#c200c2;padding:5px;margin:5px;border-radius:2px">egghead.io lesson 3: Simplifying deep updates with _produce_</summary>
    <br>
    <div style="padding:5px;">
        <iframe style="border: none;" width=760 height=427  src="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce/embed" ></iframe>
    </div>
    <a style="font-style:italic;padding:5px;margin:5px;"  href="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce">Hosted on egghead.io</a>
</details>

The Immer package exposes a default function that does all the work.

`produce(currentState, producer: (draftState) => void): nextState`

There is also a curried overload that is explained [in a later section](curried-produce).

## Example

```javascript
import produce from "immer"

const baseState = [
	{
		todo: "Learn typescript",
		done: true
	},
	{
		todo: "Try immer",
		done: false
	}
]

const nextState = produce(baseState, draftState => {
	draftState.push({todo: "Tweet about it"})
	draftState[1].done = true
})
```

The interesting thing about Immer is that the `baseState` will be untouched, but the `nextState` will reflect all changes made to `draftState`.

```javascript
// the new item is only added to the next state,
// base state is unmodified
expect(baseState.length).toBe(2)
expect(nextState.length).toBe(3)

// same for the changed 'done' prop
expect(baseState[1].done).toBe(false)
expect(nextState[1].done).toBe(true)

// unchanged data is structurally shared
expect(nextState[0]).toBe(baseState[0])
// changed data not (d√ªh)
expect(nextState[1]).not.toBe(baseState[1])
```
