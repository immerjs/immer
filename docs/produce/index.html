<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/immer/produce">
    <link rel="canonical" href="/immer/produce" />
  </head>
  <script>
    window.location.href = '/immer/produce' + window.location.search + window.location.hash;
  </script>
</html>