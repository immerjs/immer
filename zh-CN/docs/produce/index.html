<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <meta http-equiv="refresh" content="0; url=/immer/zh-CN/produce">
    <link rel="canonical" href="/immer/zh-CN/produce" />
  </head>
  <script>
    window.location.href = '/immer/zh-CN/produce' + window.location.search + window.location.hash;
  </script>
</html>