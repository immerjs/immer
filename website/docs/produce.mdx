---
id: produce
title: Using produce
---

<center>
	<div
		data-ea-publisher="immerjs"
		data-ea-type="image"
		className="horizontal bordered"
	></div>
</center> <details>
	<summary className="egghead-summary">
		egghead.io lesson 3: Simplifying deep updates with _produce_
	</summary>
	<br />
	<div>
		<iframe
			width="760"
			height="427"
			scrolling="no"
			src="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce/embed"
		></iframe>
	</div>
	<a
		className="egghead-link"
		href="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce"
	>
		Hosted on egghead.io
	</a>
</details>

The Immer package exposes a default function that does all the work.

`produce(currentState, recipe: (draftState) => void): nextState`

`produce` takes a base state, and a _recipe_ that can be used to perform all the desired mutations on the `draft` that is passed in. The interesting thing about Immer is that the `baseState` will be untouched, but the `nextState` will reflect all changes made to `draftState`.

Inside the recipe, all standard JavaScript APIs can be used on the `draft` object, including field assignments, `delete` operations, and mutating array, Map and Set operations like `push`, `pop`, `splice`, `set`, `sort`, `remove`, etc.

Any of those mutations don't have to happen at the root, but it is allowed to modify anything anywhere deep inside the draft: `draft.todos[0].tags["urgent"].author.age = 56`

Note that the recipe function itself normally doesn't return anything. However, it is possible to return in case you want to replace the `draft` object in its entirety with another object, for more details see [returning new data](return).

## Example

```javascript
import produce from "immer"

const baseState = [
	{
		title: "Learn TypeScript",
		done: true
	},
	{
		title: "Try Immer",
		done: false
	}
]

const nextState = produce(baseState, draftState => {
	draftState.push({title: "Tweet about it"})
	draftState[1].done = true
})
```

```javascript
// the new item is only added to the next state,
// base state is unmodified
expect(baseState.length).toBe(2)
expect(nextState.length).toBe(3)

// same for the changed 'done' prop
expect(baseState[1].done).toBe(false)
expect(nextState[1].done).toBe(true)

// unchanged data is structurally shared
expect(nextState[0]).toBe(baseState[0])
// ...but changed data isn't.
expect(nextState[1]).not.toBe(baseState[1])
```

### Terminology

- `(base)state`, the immutable state passed to `produce`
- `recipe`: the second argument of `produce`, that captures how the base state should be "mutated".
- `draft`: the first argument of any `recipe`, which is a proxy to the original base state that can be safely mutated.
- `producer`. A function that uses `produce` and is generally of the form `(baseState, ...arguments) => resultState`

Note that it isn't strictly necessary to name the first argument of the recipe `draft`. You can name it anything you want, for example `users`. Using `draft` as a name is just a convention to signal: "mutation is OK here".
