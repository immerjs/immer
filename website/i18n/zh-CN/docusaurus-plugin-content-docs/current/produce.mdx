---
id: produce
title: 使用 produce
---

<center>
	<div
		data-ea-publisher="immerjs"
		data-ea-type="image"
		className="horizontal bordered"
	></div>
</center>
<details>
	<summary className="egghead-summary">
		egghead.io 第 3 课：使用 produce 简化深度更新
	</summary>
	<br />
	<div>
		<iframe
			width="760"
			height="427"
			scrolling="no"
			src="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce/embed"
		></iframe>
	</div>
	<a
		className="egghead-link"
		href="https://egghead.io/lessons/javascript-simplify-deep-state-updates-using-immer-produce"
	>
		Hosted on egghead.io
	</a>
</details>

Immer 包暴露了一个完成所有工作的默认函数。

`produce(currentState, recipe: (draftState) => void): nextState`

`produce` 需要一个 `baseState`，以及一个可用于对传入的 `draft` 进行所有所需更改的 `recipe`。关于 Immer 的有趣之处在于 baseState 将保持不变，但 nextState 将反映对 DraftState 所做的所有更改.

在 `recipe` 中，所有标准的 JavaScript API 都可以在 `draft` 对象上使用，包括属性字段分配、删除操作和修改数组、Map 和 Set 操作，如 push、pop、splice、set、sort、remove 等。

这些 `mutations` 中的任何一个都不必发生在初始对象上，但它可以修改 `draft` 深处的任何内容：`draft.todos[0].tags["urgent"].author.age = 56`

请注意，`recipe` 函数通常不会返回任何内容。但是，如果您想用另一个对象完全替换 `draft`，则可以返回，有关更多详细信息，请参阅返回[新数据](return)。

## 例子

```javascript
import produce from "immer"

const baseState = [
	{
		title: "Learn TypeScript",
		done: true
	},
	{
		title: "Try Immer",
		done: false
	}
]

const nextState = produce(baseState, draftState => {
	draftState.push({title: "Tweet about it"})
	draftState[1].done = true
})
```

```javascript
// 新的 item 仅仅被添加到了 next state
// base state 没有被修改
expect(baseState.length).toBe(2)
expect(nextState.length).toBe(3)

// 同上
expect(baseState[1].done).toBe(false)
expect(nextState[1].done).toBe(true)

// 未修改的数据结构共享
expect(nextState[0]).toBe(baseState[0])
// 改变的数据不是
expect(nextState[1]).not.toBe(baseState[1])
```

### 术语

- `(base)state`， 传递给 `produce` 的不可变状态
- `recipe`: `produce` 的第二个参数，它捕获了 base state 应该如何 `mutated`。
- `draft`: 任何 `recipe` 的第一个参数，它是可以安全 `mutate` 的原始状态的代理。
- `producer`. 一个使用 `produce` 的函数，通常形式为 `(baseState, ...arguments) => resultState`

请注意，命名 `recipe` 的第一个参数 `draft` 并不是绝对必要的。您可以将其命名为任何您想要的名称，例如 `user`。使用 `draft` 作为名称只是一个约定，以表明：“这里的 `mutation` 是可以的”。
