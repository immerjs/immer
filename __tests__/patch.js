"use strict"
import produce, {setUseProxies, applyPatches} from "../src/index"

jest.setTimeout(1000)

function runPatchTest(
    base,
    producer,
    patches,
    inversePathes,
    proxyOnly = false
) {
    function runPatchTestHelper() {
        let recordedPatches
        let recordedInversePatches
        const res = produce(base, producer, (p, i) => {
            recordedPatches = p
            recordedInversePatches = i
        })

        test("produces the correct patches", () => {
            expect(recordedPatches).toEqual(patches)
            if (inversePathes)
                expect(recordedInversePatches).toEqual(inversePathes)
        })

        test("patches are replayable", () => {
            expect(applyPatches(base, recordedPatches)).toEqual(res)
        })

        test("patches can be reversed", () => {
            expect(applyPatches(res, recordedInversePatches)).toEqual(base)
        })
    }

    describe(`proxy`, () => {
        beforeAll(() => setUseProxies(true))
        runPatchTestHelper()
    })

    describe(`es5`, () => {
        if (!proxyOnly) {
            beforeAll(() => setUseProxies(false))
            runPatchTestHelper()
        }
    })
}

describe("applyPatches", () => {
    it("mutates the base state when it is a draft", () => {
        produce({a: 1}, draft => {
            const result = applyPatches(draft, [
                {op: "replace", path: ["a"], value: 2}
            ])
            expect(result).toBe(draft)
            expect(draft.a).toBe(2)
        })
    })
    it("produces a copy of the base state when not a draft", () => {
        const base = {a: 1}
        const result = applyPatches(base, [
            {op: "replace", path: ["a"], value: 2}
        ])
        expect(result).not.toBe(base)
        expect(result.a).toBe(2)
        expect(base.a).toBe(1)
    })
    it('throws when `op` is not "add", "replace", nor "remove"', () => {
        expect(() => {
            const patch = {op: "copy", from: [0], path: [1]}
            applyPatches([2], [patch])
        }).toThrowErrorMatchingSnapshot()
    })
    it("throws when `path` cannot be resolved", () => {
        // missing parent
        expect(() => {
            const patch = {op: "add", path: ["a", "b"], value: 1}
            applyPatches({}, [patch])
        }).toThrowErrorMatchingSnapshot()

        // missing grand-parent
        expect(() => {
            const patch = {op: "add", path: ["a", "b", "c"], value: 1}
            applyPatches({}, [patch])
        }).toThrowErrorMatchingSnapshot()
    })
})

describe("simple assignment - 1", () => {
    runPatchTest(
        {x: 3},
        d => {
            d.x++
        },
        [{op: "replace", path: ["x"], value: 4}]
    )
})

describe("simple assignment - 2", () => {
    runPatchTest(
        {x: {y: 4}},
        d => {
            d.x.y++
        },
        [{op: "replace", path: ["x", "y"], value: 5}]
    )
})

describe("simple assignment - 3", () => {
    runPatchTest(
        {x: [{y: 4}]},
        d => {
            d.x[0].y++
        },
        [{op: "replace", path: ["x", 0, "y"], value: 5}]
    )
})

describe("simple assignment - 4", () => {
    runPatchTest(
        new Map([["x", {y: 4}]]),
        d => {
            d.get("x").y++
        },
        [{op: "replace", path: ["x", "y"], value: 5}],
        [{op: "replace", path: ["x", "y"], value: 4}],
        true
    )
})

describe("simple assignment - 5", () => {
    runPatchTest(
        {x: new Map([["y", 4]])},
        d => {
            d.x.set("y", 5)
        },
        [{op: "replace", path: ["x", "y"], value: 5}],
        [{op: "replace", path: ["x", "y"], value: 4}],
        true
    )
})

describe("delete 1", () => {
    runPatchTest(
        {x: {y: 4}},
        d => {
            delete d.x
        },
        [{op: "remove", path: ["x"]}]
    )
})

describe("delete 2", () => {
    runPatchTest(
        new Map([["x", 1]]),
        d => {
            d.delete("x")
        },
        [{op: "remove", path: ["x"]}],
        [{op: "add", path: ["x"], value: 1}],
        true
    )
})

describe("delete 3", () => {
    runPatchTest(
        {x: new Map([["y", 1]])},
        d => {
            d.x.delete("y")
        },
        [{op: "remove", path: ["x", "y"]}],
        [{op: "add", path: ["x", "y"], value: 1}],
        true
    )
})

describe("renaming properties", () => {
    describe("nested object (no changes)", () => {
        runPatchTest(
            {a: {b: 1}},
            d => {
                d.x = d.a
                delete d.a
            },
            [
                {op: "add", path: ["x"], value: {b: 1}},
                {op: "remove", path: ["a"]}
            ]
        )
    })

    describe("nested map (no changes)", () => {
        runPatchTest(
            new Map([["a", new Map([["b", 1]])]]),
            d => {
                d.set("x", d.get("a"))
                d.delete("a")
            },
            [
                {op: "add", path: ["x"], value: new Map([["b", 1]])},
                {op: "remove", path: ["a"]}
            ],
            [
                {op: "remove", path: ["x"]},
                {op: "add", path: ["a"], value: new Map([["b", 1]])}
            ],
            true
        )
    })

    describe("nested object (with changes)", () => {
        runPatchTest(
            {a: {b: 1, c: 1}},
            d => {
                let a = d.a
                a.b = 2 // change
                delete a.c // delete
                a.y = 2 // add

                // rename
                d.x = a
                delete d.a
            },
            [
                {op: "add", path: ["x"], value: {b: 2, y: 2}},
                {op: "remove", path: ["a"]}
            ]
        )
    })

    describe("nested map (with changes)", () => {
        runPatchTest(
            new Map([["a", new Map([["b", 1], ["c", 1]])]]),
            d => {
                let a = d.get("a")
                a.set("b", 2) // change
                a.delete("c") // delete
                a.set("y", 2) // add

                // rename
                d.set("x", a)
                d.delete("a")
            },
            [
                {op: "add", path: ["x"], value: new Map([["b", 2], ["y", 2]])},
                {op: "remove", path: ["a"]}
            ],
            [
                {op: "remove", path: ["x"]},
                {op: "add", path: ["a"], value: new Map([["b", 1], ["c", 1]])}
            ],
            true
        )
    })

    describe("deeply nested object (with changes)", () => {
        runPatchTest(
            {a: {b: {c: 1, d: 1}}},
            d => {
                let b = d.a.b
                b.c = 2 // change
                delete b.d // delete
                b.y = 2 // add

                // rename
                d.a.x = b
                delete d.a.b
            },
            [
                {op: "add", path: ["a", "x"], value: {c: 2, y: 2}},
                {op: "remove", path: ["a", "b"]}
            ]
        )
    })

    describe("deeply nested map (with changes)", () => {
        runPatchTest(
            new Map([["a", new Map([["b", new Map([["c", 1], ["d", 1]])]])]]),
            d => {
                let b = d.get("a").get("b")
                b.set("c", 2) // change
                b.delete("d") // delete
                b.set("y", 2) // add

                // rename
                d.get("a").set("x", b)
                d.get("a").delete("b")
            },
            [
                {
                    op: "add",
                    path: ["a", "x"],
                    value: new Map([["c", 2], ["y", 2]])
                },
                {op: "remove", path: ["a", "b"]}
            ],
            [
                {op: "remove", path: ["a", "x"]},
                {
                    op: "add",
                    path: ["a", "b"],
                    value: new Map([["c", 1], ["d", 1]])
                }
            ],
            true
        )
    })
})

describe("minimum amount of changes", () => {
    runPatchTest(
        {x: 3, y: {a: 4}, z: 3},
        d => {
            d.y.a = 4
            d.y.b = 5
            Object.assign(d, {x: 4, y: {a: 2}})
        },
        [
            {op: "replace", path: ["x"], value: 4},
            {op: "replace", path: ["y"], value: {a: 2}}
        ]
    )
})

describe("arrays - prepend", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.unshift(4)
        },
        [{op: "add", path: ["x", 0], value: 4}]
    )
})

describe("arrays - multiple prepend", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.unshift(4)
            d.x.unshift(5)
        },
        [
            {op: "add", path: ["x", 0], value: 5},
            {op: "add", path: ["x", 1], value: 4}
        ]
    )
})

describe("arrays - splice middle", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.splice(1, 1)
        },
        [{op: "remove", path: ["x", 1]}]
    )
})

describe("arrays - multiple splice", () => {
    runPatchTest(
        [0, 1, 2, 3, 4, 5, 0],
        d => {
            d.splice(4, 2, 3)
            d.splice(1, 2, 3)
        },
        [
            {op: "replace", path: [1], value: 3},
            {op: "replace", path: [2], value: 3},
            {op: "remove", path: [5]},
            {op: "remove", path: [4]}
        ]
    )
})

describe("arrays - modify and shrink", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x[0] = 4
            d.x.length = 2
        },
        [
            {op: "replace", path: ["x", 0], value: 4},
            {op: "replace", path: ["x", "length"], value: 2}
        ]
    )
})

describe("arrays - prepend then splice middle", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.unshift(4)
            d.x.splice(2, 1)
        },
        [
            {op: "replace", path: ["x", 0], value: 4},
            {op: "replace", path: ["x", 1], value: 1}
        ]
    )
})

describe("arrays - splice middle then prepend", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.splice(1, 1)
            d.x.unshift(4)
        },
        [
            {op: "replace", path: ["x", 0], value: 4},
            {op: "replace", path: ["x", 1], value: 1}
        ]
    )
})

describe("arrays - truncate", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.length -= 2
        },
        [{op: "replace", path: ["x", "length"], value: 1}],
        [
            {op: "add", path: ["x", 1], value: 2},
            {op: "add", path: ["x", 2], value: 3}
        ]
    )
})

describe("arrays - pop twice", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.pop()
            d.x.pop()
        },
        [{op: "replace", path: ["x", "length"], value: 1}]
    )
})

describe("arrays - push multiple", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.push(4, 5)
        },
        [
            {op: "add", path: ["x", 3], value: 4},
            {op: "add", path: ["x", 4], value: 5}
        ],
        [{op: "replace", path: ["x", "length"], value: 3}]
    )
})

describe("arrays - splice (expand)", () => {
    runPatchTest(
        {x: [1, 2, 3]},
        d => {
            d.x.splice(1, 1, 4, 5, 6)
        },
        [
            {op: "replace", path: ["x", 1], value: 4},
            {op: "add", path: ["x", 2], value: 5},
            {op: "add", path: ["x", 3], value: 6}
        ],
        [
            {op: "replace", path: ["x", 1], value: 2},
            {op: "remove", path: ["x", 3]},
            {op: "remove", path: ["x", 2]}
        ]
    )
})

describe("arrays - splice (shrink)", () => {
    runPatchTest(
        {x: [1, 2, 3, 4, 5]},
        d => {
            d.x.splice(1, 3, 6)
        },
        [
            {op: "replace", path: ["x", 1], value: 6},
            {op: "remove", path: ["x", 3]},
            {op: "remove", path: ["x", 2]}
        ],
        [
            {op: "replace", path: ["x", 1], value: 2},
            {op: "add", path: ["x", 2], value: 3},
            {op: "add", path: ["x", 3], value: 4}
        ]
    )
})

describe("simple replacement", () => {
    runPatchTest({x: 3}, _d => 4, [{op: "replace", path: [], value: 4}])
})

describe("same value replacement - 1", () => {
    runPatchTest(
        {x: {y: 3}},
        d => {
            const a = d.x
            d.x = a
        },
        []
    )
})

describe("same value replacement - 2", () => {
    runPatchTest(
        {x: {y: 3}},
        d => {
            const a = d.x
            d.x = 4
            d.x = a
        },
        []
    )
})

describe("same value replacement - 3", () => {
    runPatchTest(
        {x: 3},
        d => {
            d.x = 3
        },
        []
    )
})

describe("same value replacement - 4", () => {
    runPatchTest(
        {x: 3},
        d => {
            d.x = 4
            d.x = 3
        },
        []
    )
})

describe("same value replacement - 5", () => {
    runPatchTest(
        new Map([["x", 3]]),
        d => {
            d.set("x", 4)
            d.set("x", 3)
        },
        [],
        [],
        true
    )
})

describe("simple delete", () => {
    runPatchTest(
        {x: 2},
        d => {
            delete d.x
        },
        [
            {
                op: "remove",
                path: ["x"]
            }
        ]
    )
})
