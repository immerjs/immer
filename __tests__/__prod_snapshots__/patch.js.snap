// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`applyPatches > throws when \`op\` is not "add", "replace", nor "remove" 1`] = `[Error: [Immer] minified error nr: 17. Full error at: https://bit.ly/3cXEKWf]`;

exports[`applyPatches > throws when \`path\` cannot be resolved 1`] = `[Error: [Immer] minified error nr: 18. Full error at: https://bit.ly/3cXEKWf]`;

exports[`applyPatches > throws when \`path\` cannot be resolved 2`] = `[Error: [Immer] minified error nr: 18. Full error at: https://bit.ly/3cXEKWf]`;
